netket/utils/group/_permutation_group.py

```diff
@@ -161,23 +161,16 @@ def product_table(self) -> Array:
             perms = self.to_array()
             inverse = perms[self.inverse].squeeze()
             n_symm = len(perms)
-            product_table = np.zeros([n_symm, n_symm], dtype=int)
-
-            inv_t = inverse.transpose()
-            perms_t = perms.transpose()
-            inv_elements = perms_t[inv_t].reshape(-1, n_symm * n_symm).transpose()
-
-            inv_perms = [HashableArray(element) for element in inv_elements]
+            product_table = []
```

Comment:
maybe you could preallocate it and then modify it in-place?

Reply:
That only pushes applicability a little bit further. For a 20x20 square lattice, you have 4G entries, for 30x30, 47G. At this point, the calculation is not limited by the overheads of a few-thousand-iteration loop

Reply:
it's not really for performance reasons (though it would help in a tiny way).

It's really just to have a more portable code. Numpy-like code code can be easily moved to numba/jax if needs ever be.
Pure python can't.

It's also more concise. Probably it's useless, but If we ever will need to move the code around it will make life easier.

I'd do something like
```python
product_table = np.zeros([n_symm, n_symm], dtype=int)
for (i,g_int) in enumerate(inverse):
   row_perms = perms[:, g_inv]
   product_table[i,:] = np.vectorize(lambda perm: lookup[HashableArray(perm)])(row_perms)

```

Reply:
Aaah I see, I was confused by the old code snippet github shows here...

BTW is list comprehension/`np.vectorize` any better than a plain old loop? Because
```python
product_table = np.zeros([n_symm, n_symm], dtype=int)
for (i,g_int) in enumerate(inverse):
   row_perms = perms[:, g_inv]
   for j,perm in enumerate(perms):
      product_table[i,j] = lookup[HashableArray(perm)]
```
is more transparent imo

Reply:
frankly I like your snippet better than mine (and much more than what is currently in netket, which is a bit opaque IMO).

But I fear the nested loop will slow it down a lot. 
I was using `np.vectorize` to strike a balance.

Reply:
> I was using `np.vectorize` to strike a balance.

According to the numpy docs `vectorize` is no more performant than a `for` loop.

> But I fear the nested loop will slow it down a lot.

My understanding is that every call to a loop body adds some constant overhead to the runtime, so it doesn't matter if you have one N*N-iteration loop or an N-iteration loop with an N-iteration loop in the body.

I've run some quick benchmarks, finding the product table of `nk.graph.Square(12).space_group()` took 36s with the old code, 34s with my snippet above, and 44s with your snippet. It's weird that `vectorize` is _slower,_ but such is life. I'll commit my version in a sec